### ISSUE TITLE
crash when cancelling auto-pilot

### ISSUE THREAD

**Issue by comnom** [endless-sky/endless-sky#4237]
Cancelling auto-pilot via ```L``` in an unihabited system with a travel destination selected causes an access violation.

The stack right before the access violation:
<details>
<code>
0x00000000005187e0 in StellarObject::GetPlanet (this=0x0) at StellarObject.cpp:62
0x00000000004163e7 in AI::MovePlayer (this=0x31b91250, ship=..., player=...) at AI.cpp:3198
0x000000000040567d in AI::Step (this=0x31b91250, player=...) at AI.cpp:470
0x000000000044c5a4 in Engine::CalculateStep (this=0x31b91050) at Engine.cpp:1190
0x000000000044c3f6 in Engine::ThreadEntryPoint (this=0x31b91050) at Engine.cpp:1165
</code>
</details>

Here's a save. To replicate, tap L right after begin to jump from Prakacha'a:
[Zachita_XenDrik.txt](https://github.com/endless-sky/endless-sky/files/3000695/Zachita_XenDrik.txt)

Original report from Zachita on discord:
![Screen_Shot_2019-03-21_at_7 37 19_PM](https://user-images.githubusercontent.com/22107649/54883813-193fdd00-4e40-11e9-90ec-1503eb2f3639.png)
Essentially, when I start jumping from the uninhabited system Prakacha, to the Unfettered system, my game will crash/close if I hit L prior to entering the unfettered system.

If I hit L while in the jump animation from the inhabited system of Sich'ka'ara, the game does not crash, but if I do it from the uninhabited system of Prakacha'a, it crashes.

**Comment by jostephd** [comment]
I get a segfault if I tap L _twice_ after the jump animation from Prakacha'a has started playing.

Before the segfault it prints "Landing on Firelode" at the bottom of the screen. Firelode is in Ehma Ti, the system being jumped to.

**Comment by gunqqer** [comment]
Cannot replicate here on KDE Neon (Ubuntu 18.04 base) on commit bdded151 using either the given save file. What version are you guys using that is getting the crash?

**Comment by jostephd** [comment]
Commit 3c61ba2c941634fa7179523b7e37b99a5f803302 using the given save file.

**Comment by MCOfficer** [comment]
Can repro on both 3c61ba2c941634fa7179523b7e37b99a5f803302 and bdded151a513a0a115f16cf5cd5d25cf92ca8329. Windows 10 64bit.

@gunqqer just to be clear: you want to jump to the uninhabited system, then wait for the jump animation, and that's when you spam the L key. the timing is slightly dodgy, if you press too early the jump will cancel. I'm not sure if the jump has already been initiated when this happens, perhaps the ship was still maneuvering into position.

**Comment by jostephd** [comment]
You have wait for the "jumped into the system" animation to finish, then when the "jumping out of the system" animation starts, press L twice.

**Comment by gunqqer** [comment]
Ok was misunderstanding how to reproduce, now can. Ref Video: https://youtu.be/MzQTViPISqw
Full segfault from `dmesg`
`
[421689.393918] endless-sky[29727]: segfault at 60 ip 0000560cf987ae50 sp 00007f0664645e28 error 4 in endless-sky[560cf9670000+26c000]

[421689.393929] Code: c8 f2 0f 10 05 91 1a 01 00 f2 0f 59 c1 74 c4 f2 0f 10 0d 6b 39 01 00 48 83 c4 10 5b f2 0f 5d c8 66 0f 28 c1 c3 90 0f 1f 40 00 <48> 8b 47 60 c3 90 66 2e 0f 1f 84 00 00 00 00 00 53 0f b6 05 58 92
`

**Comment by MCOfficer** [comment]
I found the issue, but i'm not sure about the best way to fix it.

In hindsight, it's pretty obvious:
https://github.com/endless-sky/endless-sky/blob/bdded151a513a0a115f16cf5cd5d25cf92ca8329/source/AI.cpp#L3168-L3183
this creates the variable `next` (which will be used as the next target), and then finds a StellarObject that fits it ( = one you can land on). Since there are no planets to land on in Prakacha'a, at the end of this segment, `next` still points at 0x0.

https://github.com/endless-sky/endless-sky/blob/bdded151a513a0a115f16cf5cd5d25cf92ca8329/source/AI.cpp#L3184-L3194
This part is supposed to resolve that issue, but once again, it only checks for possibly landable objects.

https://github.com/endless-sky/endless-sky/blob/bdded151a513a0a115f16cf5cd5d25cf92ca8329/source/AI.cpp#L3184-L3194
finally, the segfault happens at L3197, because `next` is still pointing at 0x0. Either way, it would probably segfault once the ship tries to steer to the target set at L3195.

<details>
 <summary>here's the full callstack right at the segfault</summary>

```#0 0x51cc80	StellarObject::GetPlanet(this=0x0) (C:\Users\Florian\Documents\GitHub\endless-sky\source\StellarObject.cpp:62)
#1 0x417128	AI::MovePlayer(this=0xfa24470, ship=..., player=...) (C:\Users\Florian\Documents\GitHub\endless-sky\source\AI.cpp:3197)
#2 0x405790	AI::Step(this=0xfa24470, player=...) (C:\Users\Florian\Documents\GitHub\endless-sky\source\AI.cpp:470)
#3 0x44de68	Engine::CalculateStep(this=0xfa24270) (C:\Users\Florian\Documents\GitHub\endless-sky\source\Engine.cpp:1190)
#4 0x44dcb7	Engine::ThreadEntryPoint(this=0xfa24270) (C:\Users\Florian\Documents\GitHub\endless-sky\source\Engine.cpp:1165)
#5 0x6c7df9	std::__invoke_impl<void, void (Engine::* const&)() (D:/mingw64-6.3.0-rev1/lib/gcc/x86_64-w64-mingw32/6.3.0/include/c++/functional:227)
#6 0x6e6857	std::__invoke<void (Engine::* const&)() (D:/mingw64-6.3.0-rev1/lib/gcc/x86_64-w64-mingw32/6.3.0/include/c++/functional:251)
#7 0x56f49b	std::_Mem_fn_base<void (Engine::*)() (D:/mingw64-6.3.0-rev1/lib/gcc/x86_64-w64-mingw32/6.3.0/include/c++/functional:604)
#8 0x5a1ceb	std::_Bind_simple<std::_Mem_fn<void (Engine::*)() (D:/mingw64-6.3.0-rev1/lib/gcc/x86_64-w64-mingw32/6.3.0/include/c++/functional:1391)
#9 0x5a1d97	std::_Bind_simple<std::_Mem_fn<void (Engine::*)() (D:/mingw64-6.3.0-rev1/lib/gcc/x86_64-w64-mingw32/6.3.0/include/c++/functional:1380)
#10 0x5f9cdc	std::thread::_State_impl<std::_Bind_simple<std::_Mem_fn<void (Engine::*)() (D:/mingw64-6.3.0-rev1/lib/gcc/x86_64-w64-mingw32/6.3.0/include/c++/thread:196)
#11 0x6fd0731f	libstdc++-6!atomic_flag_test_and_set_explicit() (D:\mingw64-6.3.0-rev1\bin\libstdc++-6.dll:??)
#12 0x64944dca	pthread_create_wrapper() (C:\Users\Florian\Documents\GitHub\endless-sky\libwinpthread-1.dll:??)
#13 0x7ff9d73aaa96	msvcrt!_beginthreadex() (C:\WINDOWS\System32\msvcrt.dll:??)
#14 0x7ff9d73aab6c	msvcrt!_endthreadex() (C:\WINDOWS\System32\msvcrt.dll:??)
#15 0x7ff9d6283dc4	KERNEL32!BaseThreadInitThunk() (C:\WINDOWS\System32\kernel32.dll:??)
#16 0x7ff9d74e3691	ntdll!RtlUserThreadStart() (C:\WINDOWS\SYSTEM32\ntdll.dll:??)
#17 ??	?? () (??:??)
</details>

**Comment by jostephd** [comment]
The condition in that `if` looks like it's comparing the ship's position in the current system with the target planet position in the target planet's system, so maybe this?

```
diff --git a/source/AI.cpp b/source/AI.cpp
index 825d3334..063d03cc 100644
--- a/source/AI.cpp
+++ b/source/AI.cpp
@@ -3156,7 +3156,8 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player)
 		if(!message.empty())
 			Audio::Play(Audio::Get("fail"));
 		
-		const StellarObject *target = ship.GetTargetStellar();
+		// Get the target planet, if it's in the current system.
+		const StellarObject *target = ship.GetTargetSystem() ? ship.GetTargetStellar() : nullptr;
 		if(target && (ship.Position().Distance(target->Position()) < target->Radius() || ship.Zoom() < 1.))
 		{
 			// Special case: if there are two planets in system and you have one
```

>  finally, the segfault happens at L3197, because `next` is still pointing at 0x0. Either way, it would probably segfault once the ship tries to steer to the target set at L3195.

I suppose we could add a null check before dereferencing it?

**Comment by MCOfficer** [comment]
that is your expertise. i'd just throw it all into a null check and call it a day, but the game is so complex that the simplest solution might not be the best...

**Comment by tehhowch** [comment]
Should be addressed by c095f8e33250942b0f28dd1559cc45f133ad1ff0
